File: sdk/typescript/src/tools/scope.ts
export class FileScopePolicy {
  private agentScopes: Map<string, string[]> = new Map();

  setScope(agentID: string, paths: string[]): void {
    this.agentScopes.set(agentID, paths);
  }

  getScope(agentID: string): string[] | undefined {
    return this.agentScopes.get(agentID);
  }

  removeScope(agentID: string): void {
    this.agentScopes.delete(agentID);
  }

  allow(agentID: string, resource: string): boolean {
    const scope = this.agentScopes.get(agentID);
    if (!scope) return false;
    if (scope.length === 0) return true;

    const absPath = normalizePath(resource);
    for (const s of scope) {
      const scopeAbs = normalizePath(s);
      if (absPath === scopeAbs || absPath.startsWith(scopeAbs + '/')) {
        return true;
      }
    }
    return false;
  }

  validate(agentScopes: Map<string, string[]>): Error | null {
    let globalCount = 0;
    const scopes: string[][] = [];

    for (const [, scope] of agentScopes) {
      if (scope.length === 0) globalCount++;
      scopes.push(scope);
    }

    if (globalCount > 1) {
      return new Error(`同一批次中最多允许 1 个任务拥有全局写权限，当前有 ${globalCount} 个`);
    }

    for (let i = 0; i < scopes.length; i++) {
      if (scopes[i].length === 0) continue;
      for (let j = i + 1; j < scopes.length; j++) {
        if (scopes[j].length === 0) continue;
        const overlap = findScopeOverlap(scopes[i], scopes[j]);
        if (overlap) {
          return new Error(`任务 ${i + 1} 和任务 ${j + 1} 的 scope 存在重叠: ${overlap}`);
        }
      }
    }
    return null;
  }
}

function normalizePath(p: string): string {
  return p.replace(/\\/g, '/').replace(/\/+/g, '/').replace(/\/$/, '');
}

function findScopeOverlap(scopeA: string[], scopeB: string[]): string | null {
  for (const a of scopeA) {
    const cleanA = normalizePath(a);
    for (const b of scopeB) {
      const cleanB = normalizePath(b);
      if (cleanA === cleanB) return cleanA;
      if (cleanA.startsWith(cleanB + '/')) return `${cleanA} (属于 ${cleanB})`;
      if (cleanB.startsWith(cleanA + '/')) return `${cleanB} (属于 ${cleanA})`;
    }
  }
  return null;
}

File: agentprimordia/internal/tools/scope.go
package tools

import (
	"fmt"
	"path/filepath"
	"strings"
	"sync"

	"agentprimordia/internal/concurrency"
)

type ScopePolicy interface {
	Allow(agentID, resource string) bool
	Validate(agentScopes map[string][]string) error
}

type FileScopePolicy struct {
	mu          sync.RWMutex
	agentScopes map[string][]string
}

func NewFileScopePolicy() *FileScopePolicy {
	return &FileScopePolicy{
		agentScopes: make(map[string][]string),
	}
}

func (p *FileScopePolicy) SetScope(agentID string, paths []string) {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.agentScopes[agentID] = paths
}

func (p *FileScopePolicy) GetScope(agentID string) []string {
	p.mu.RLock()
	defer p.mu.RUnlock()
	paths := p.agentScopes[agentID]
	if paths == nil {
		return nil
	}
	cp := make([]string, len(paths))
	copy(cp, paths)
	return cp
}

func (p *FileScopePolicy) RemoveScope(agentID string) {
	p.mu.Lock()
	defer p.mu.Unlock()
	delete(p.agentScopes, agentID)
}

func (p *FileScopePolicy) Allow(agentID, resource string) bool {
	p.mu.RLock()
	defer p.mu.RUnlock()

	scope, exists := p.agentScopes[agentID]
	if !exists {
		// 未注册的 Agent 默认access denied
		return false
	}

	if len(scope) == 0 {
		// 已注册但未设置路径限制的 Agent 默认access denied
		// 如需全局权限，请显式设置 scope 为 ["/"] 或具体路径
		return false
	}

	absPath := filepath.Clean(resource)
	for _, s := range scope {
		scopeAbs := filepath.Clean(s)
		// 根路径 "/" 或 "\" 匹配所有路径（Windows 兼容）
		if scopeAbs == "/" || scopeAbs == `\` {
			return true
		}
		if absPath == scopeAbs || strings.HasPrefix(absPath, scopeAbs+string(filepath.Separator)) {
			return true
		}
	}

	return false
}

func (p *FileScopePolicy) Validate(agentScopes map[string][]string) error {
	scopes := make([][]string, 0, len(agentScopes))
	for _, scope := range agentScopes {
		scopes = append(scopes, scope)
	}
	return concurrency.ValidateScopes(scopes)
}

func (p *FileScopePolicy) ValidateCurrent() error {
	p.mu.RLock()
	defer p.mu.RUnlock()

	scopes := make([][]string, 0, len(p.agentScopes))
	for _, scope := range p.agentScopes {
		scopes = append(scopes, scope)
	}
	return concurrency.ValidateScopes(scopes)
}

func NewScopeDeniedError(agentID, resource string) error {
	return fmt.Errorf("agent %s denied access to resource %s", agentID, resource)
}
